
# **Promiseチェーン**

`.then（）`および `.catch（）`メソッドは常にpromiseを返すため、複数の `.then（）`呼び出しを連結できます。

最初に、promiseを返す遅延関数を作成します。
返されたpromiseは、指定された秒数後に解決します。

```js
const delay = ms => new Promise(resolve => setTimeout(resolve, ms));
```

このスニペットでは、すぐに実行されないように、関数を使用してプロミスをラップしています。

```js
const delay = ms => new Promise(resolve => setTimeout(resolve, ms));

delay(2000).then(() => console.log("Resolved after 2 seconds"));
```

`.then（）`コールバックのステートメントは、2秒後に `delay`が解決された後にのみ実行されます。

複数の `.then（）`呼び出しをチェーンする方法は次のとおりです。


```js
const delay = ms => new Promise(resolve => setTimeout(resolve, ms));

delay(2000)
  .then(() => {
    console.log("Resolved after 3 seconds");
    return delay(3000);
  })
  .then(() => {
    console.log("Resolved after 2 seconds");
    return delay(2000);
  })
  .then(() => {
    console.log("Resolved after 1 seconds");
    throw new Error();
  })
  .catch(() => console.log("An error has occurred..."))
  .then(() => console.log("Finished"));
```

これは2秒後に解決してpromiseを返し、さらに3秒後に解決して新しいpromiseを返します。
これは、最後のpromiseがエラーをスローするまで起こります。エラーは`catch`でキャッチされ、最後の`then` チェーンで終了します。
